---
title: "BCB420 - Computational Systems Biology"
author: "Juan De Los Rios"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: '2'
    df_print: paged
  html_notebook:
    toc: yes
    toc_depth: 2
subtitle: Assignment 3 - Dataset Pathway and Network Analysis
bibliography: A3_DeLosRios.bib
---
```{r main, child = 'A2_DeLosRios.Rmd'}
```
# Required packages. 
We will be using functions from the knitr [@R-knitr] package. We will also be using code taken from BCB420 lectures. [@Isserlin2022]
```{r} 
if (!requireNamespace("knitr", quietly = TRUE))
  install.packages("knitr")
```

# Non-thresholded Enrichment Analysis. 
Having performed a thresholded overrepresentation analysis, we will now perform a non-thresholded enrichment analysis using a ranked list of our differentially expressed genes and compare it with the results of our ORA to see if it grants us any insight into the impact of hypolglycaemia in human primary astrocytes.  

## Preparing Data for GSEA. 
For our enrichment analysis we will be using the Preranked analysis on the GSEA application [@subramanian2005gene] version 4.2.3. We utilize a Preranked analysis as opposed to a standard GSEA since it is better suited to summarizing the overall trends in gene expression within our list whereas a standard GSEA would be better suited for a more thorough look into gene expression within each phenotype. For our geneset database we will use the March 1, 2022 files procured from the baderlab collection containing GO biological process, no IEA and pathways. This database focuses on the categories we're most interested in investigating, biological pathways impacted by hypoglycaemia, and its similarity to the categories we chose during our ORA makes comparing their results simpler.  

To start off, we will created a ranked list of our differentially expressed genes. The .rnk file we create will be the input we give GSEA so it can perform its analysis.  
```{r}
GeneRankedList <- data.frame(GeneName = qlf_merged_frames$HUGO_symbol, 
                             rank = qlf_merged_frames$qlf_log10_pvalue * 
                                    sign(qlf_merged_frames$qlf_logFC))
GeneRankedList <- GeneRankedList[order(GeneRankedList$rank, decreasing = TRUE),]
write.table(GeneRankedList, file = "GeneRankedList.rnk", sep = "\t", 
            quote = FALSE, row.names = FALSE)
```
## GSEA Results. 
We summarize our results by displaying the top 10 pathways returned for both over- and underexpressed genes.  
```{r}
GSEAResults <- read.table(file = file.path(getwd(), "gsea_report_for_na_pos_1648080297765.tsv"), sep = "\t", header = TRUE, stringsAsFactors = FALSE, check.names = FALSE, fill = TRUE)
knitr::kable(data.frame(GSEAResults[1:10, c(1, 4:8)]), format = "html", digits = 5)

GSEAResults <- read.table(file = file.path(getwd(), "gsea_report_for_na_neg_1648080297765.tsv"), sep = "\t", header = TRUE, stringsAsFactors = FALSE, check.names = FALSE, fill = TRUE)
knitr::kable(data.frame(GSEAResults[1:10, c(1, 4:8)]), format = "html", digits = 5)
```
                       
There are few direct comparisons we can make between the results of our GSEA and the results of our ORA. Both feature pathways related to ATP synthesis and mitochondrial activity as top hits for overexpressed genes, but in terms of underexpressed genes share 0 top hits. The mitochondrial stress pathways we focused on in our discussion of results also return in the hits for overexpressed genes, although many of them would be filtered out if we used 0.05 as a threshold value for p-values and FDR.   
While pathways that make an appearance in the results of both analyses are likely candidates for further exploration into how they are impacted by hypoglycaemia, it is difficult to say whether pathways that show up in one analysis but not the other should be excluded or whether any one analysis is measurably worse for excluding certain pathways. Thus, our comparison of results cannot proceed much further beyond keeping in mind these shared top hits as we proceed with our interpretation of our results.

# Cytoscape Visualization.   
In order to better interpret the results of our enrichment analysis, we will visualize them using the Cytoscape app [@shannon2003cytoscape]. In addition, we will be employing the EnrichmentMap add-on for Cytoscape [@merico2010enrichment] that will create a network out of our results for us, facilitating the analysis of broader trends evidenced by our dataset.   

![Network Prior to Manual Layout](`r getwd()`/A3GSEANetwork.jpeg)   
**FIGURE 1. Enrichment Map Network Prior to Manual Layout** The network generated by EnrichmentMap prior to any modifications in terms of layout. There are 421 nodes, each representing an enriched pathway and the ranked genes within it from our analysis, and 1290 edges, representing the links between each pathway based on gene overlap. For a pathway to appear in our visualization, it must have met the thresholds established by an FDR q-value of 1 and a combined Jaccard-Overlap value of 0.375. Node size is determined by FDR q-value; the smaller the q-value associated with the node, the larger the node. Node color is determined by its NES column, with red denoting networks where genes were significantly overexpressed and bluye denoting networks where genes were significantly underexpressed.   

![Network After Manual Layout](`r getwd()`/Data Set 1.jpeg)   
**FIGURE 2. Enrichment Map Network After Manual Layout** Our enrichment map network after being modified for the sake of enabling further interpretation and analysis. We used the yFiles organic layout as the basis of our modified network, at which point clusters of related pathways which were deemed too cluttered to visually interpret were reorganized using the Circular layout. Finally, some nodes were manually moved to remove areas where multiple nodes were overlapping.  

![Network After Manual Layout](`r getwd()`/CollapsedNetwork.jpeg)         
**FIGURE 3. Enrichment Map Theme Network** Our enrichment map after its many pathways were collapsed into broader categories to remove redundancies. Blue and red nodes correspond to pathways which were not linked to any others and thus were not collapsed in the creation of the theme network.   

![Network After Manual Layout](`r getwd()`/AnnotatedPathway.jpeg)    
**FIGURE 4. Annotation of 'ATF4 activates genes in response to endoplasmic reticulum - homo sapiens' Pathway** Annotation of a pathway of interest obtained from WikiPathways. Our gene ranked list was uploaded onto Cytoscape and used to determine the color of each gene in the pathway, with red denoting genes that were significantly overexpressed to varying degrees and white denoting genes that were not significantly expressed.   

# Interpretation of Results. 
Once again, the purpose of our enrichment analysis and visualization of results was to see if we could confirm or otherwise support the findings of the paper from which we downloaded our initial data set. Their findings were that endoplasmic-reticulum stress-related genes involved in the unfolded protein response such as TXNIP, HSPA5, XBP1, and MANF were significantly overexpressed following low glucose exposure but underexpressed when subjected to recurrent low glucose. [@potter2020analysis] Our GSEA analysis results demonstrated the presence of enriched pathways relating to endoplasmic reticulum stress and unfolded protein response, most significantly the overexpression of ER stress genes activated by ATF4, in our dataset. However, most of these pathways were associated with FDR q-values larger than we could comfortable concluded pointed to significant enrichment.   
Further analysis achieved by visualizing our results helps address this lack of enrichment. Our enrichment map confirmed the ATF4 pathway to be significantly enriched in our dataset, and investigating it in more detail granted us insight into precisely how this pathway reacts to hypoglycaemia. ATF4, the transcription factor responsible for responding to glucose deprivation, is capable of inducing significant expression in HERPUD1, a gene involved in the unfolded protein response, without achieving substantial expression itself. Thus, even though endoplasmic-reticulum stress-related genes involved in the unfolded protein response may be significantly expressed as our paper predicted, it does not necessitate the significant expression of related transcription factors and thus results in the corresponding gene pathways being less enriched than we initially expected.    

Our GSEA analysis similar reconfirms the findings from another paper on the impact of hypoglycaemia on HPAs, which concluded that in order to maintain cell ATP levels mitochondrial genes are overexpressed as the mitochondria reorients itself to rely on fatty acid oxidation as its main source of energy in the absence of glucose. [@potter2018human] Our visualization of our GSEA results and subsequent collapse of the network in themes shows that a great deal of pathways relating to electron coupled ATP synthesis are overexpressed during hypoglycaemia while pathways relating to glycolysis are naturally underexpressed.   

# Link to Jounral. 
https://github.com/bcb420-2022/Juan_De_Los_Rios/wiki/Assignment3.  
